File: src/main/kotlin/k/strategy/WinningStrategy.kt
package k.strategy

import java.security.SecureRandom

class WinningStrategy(private val won: Boolean = false, private val prevHand: Hand? = null) :
    Strategy {
  private val random = SecureRandom()

  override fun nextHand(): Pair<Strategy, Hand> {
    if (!won) {
      val ph = Hand.getHand(random.nextInt(3))
      val st = WinningStrategy(won, ph)
      return Pair(st, ph)
    }
    return Pair(this, prevHand!!)
  }

  override fun study(win: Boolean): Strategy {
    return WinningStrategy(win, prevHand)
  }
}

File: src/main/scala/s/strategy/delegate/WinningStrategy.scala
package s.strategy.delegate

import s.strategy.Hand

import java.security.SecureRandom

final case class WinningStrategy(
    won: Boolean,
    prevHand: Option[Hand]
) extends Strategy {
  val random = new SecureRandom()

  def this() = {
    this(false, None)
  }

  override def nextHand: (Strategy, Hand) = {
    if (!won) {
      val ph = Hand.getHand(random.nextInt(3))
      val st = WinningStrategy(won, Some(ph))
      (st, ph)
    } else (this, prevHand.get)
  }

  override def study(win: Boolean): WinningStrategy =
    WinningStrategy(win, prevHand)
}
